### PR TITLE
routing: add outgoing channel restriction

### PR DESCRIPTION
**Pull request by joostjager**
This PR adds the ability to SendPayment to restrict the payment to a specific outgoing channel. 

One of the use cases of this is to be able to balance channels passively. Paying a higher fee for a payment, but at the same time increasing the receive capacity of a specific channel.

This is an alternative to making a dedicated payment to yourself to achieve the same goal, which may overall be more costly in terms of routing fees.

### CODE REVIEW

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -1951,6 +1951,11 @@ var sendPaymentCommand = cli.Command{
 			Name:  "final_cltv_delta",
 			Usage: "the number of blocks the last hop has to reveal the preimage",
 		},
+		cli.Uint64Flag{
+			Name:  "outgoing_chan_id",
+			Usage: "the 8-byte compact channel id to use for payment",
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: for consistency, should we refer to it as a short channel id? here and below?

File: lnrpc/rpc.proto [hunk 2]
```diff
@@ -774,6 +774,10 @@ message SendRequest {
     send the payment.
     */
     FeeLimit fee_limit = 8;
+
+    // outgoingChannelID is the channel that needs to be taken to the first
```
Comment (cfromknecht) [inline lnrpc/rpc.proto]: multi-line comments should use /** i think

File: routing/pathfind_test.go [hunk 3]
```diff
@@ -1918,3 +1918,94 @@ func TestNewRouteFromEmptyHops(t *testing.T) {
 		t.Fatalf("expected empty hops error: instead got: %v", err)
 	}
 }
+
+// TestRestrictOutgoingChannel asserts that a outgoing channel restriction is
+// obeyed by the path finding algorithm.
+func TestRestrictOutgoingChannel(t *testing.T) {
```
Comment (cfromknecht) [inline routing/pathfind_test.go]: great test 👍

**Comment by joostjager** [comment]
@cfromknecht ptal

File: cmd/lncli/commands.go [hunk 4]
```diff
@@ -1951,6 +1951,11 @@ var sendPaymentCommand = cli.Command{
 			Name:  "final_cltv_delta",
 			Usage: "the number of blocks the last hop has to reveal the preimage",
 		},
+		cli.Uint64Flag{
+			Name:  "outgoing_chan_id",
+			Usage: "short channel id of the outgoing channel to use for payment",
```
Comment (wpaulino) [inline cmd/lncli/commands.go]: Nit: should note this is for the first hop.

File: cmd/lncli/commands.go [hunk 5]
```diff
@@ -2186,6 +2192,12 @@ var payInvoiceCommand = cli.Command{
 			Usage: "percentage of the payment's amount used as the" +
 				"maximum fee allowed when sending the payment",
 		},
+		cli.Uint64Flag{
+			Name: "outgoing_chan_id",
+			Usage: "short channel id of the outgoing channel to " +
```
Comment (wpaulino) [inline cmd/lncli/commands.go]: Same here.

File: routing/pathfind_test.go [hunk 6]
```diff
@@ -1918,3 +1918,94 @@ func TestNewRouteFromEmptyHops(t *testing.T) {
...
+			outgoingChannelID: &outgoingChannelID,
+		},
+		sourceNode, target, paymentAmt,
+	)
+	if err != nil {
+		t.Fatalf("unable to find path: %v", err)
+	}
+	route, err := newRoute(
+		paymentAmt, infinity, sourceVertex, path, startingHeight,
+		finalHopCLTV)
```
Comment (wpaulino) [inline routing/pathfind_test.go]: Nit: closing paren should go on newline.

**Comment by joostjager** [comment]
@wpaulino ptal

**Pull request merged**